Adding external live preview

heloess opened the pull request:
Added external live preview button that opens a HTML file on browser tab which shows intermediates images and final image sequentially while generating. When save intermediates image and also save final tick button is active, they also save images temporarily to scripts folder as saved.png for HTML file to show them.

Instead of sticking to the other two tick buttons, I attempted add a new button as "Enable external preview" that only temporarily save the images as saved.png to the scripts folder, but I couldn't figure it out. It would be great if someone could add this.


https://github.com/AlUlkesh/sd_save_intermediate_images/assets/96503452/5506ab5f-1ebb-47b2-882c-c75476110cbc

heloess committed:
Update sd_save_intermediate_images.py

Saving intermediate and final image to Scripts folder temporary and adding button to open image viewer HTML file

heloess committed:
Add files via upload

Image viewer HTML file for external live preview.

heloess committed:
Deleting unnecessary import

heloess committed:
Update sd_save_intermediate_images.py

Edited def handle_image_saved

AlUlkesh commented:
Hi, very nice idea. Thank you.

A few remarks:

1. I think the new button should be inside the extension's accordion:

![image](https://github.com/AlUlkesh/sd_save_intermediate_images/assets/99896447/eb0abc9b-15c3-4743-8b6e-748c14338063)

2. The new button is active at all times. So if you press it and there is no saved.png yet, strange things happen.

3. Wouldn't it make handling easier if the Viewer is openend in a new window instead of a new tab?

4. I'd prefer it, if the html and the saved.png are in a new subdirectory. For example:

`os.path.join(scripts.basedir(), "viewer", "saved.png")`

> Instead of sticking to the other two tick buttons, I attempted add a new button as "Enable external preview" that only temporarily save the images as saved.png to the scripts folder, but I couldn't figure it out. It would be great if someone could add this.

Wouldn't it work, if you take this out of the IF block?

![image](https://github.com/AlUlkesh/sd_save_intermediate_images/assets/99896447/45b4499f-a581-47d0-9c8b-e0d806333faa)

heloess committed:
Update sd_save_intermediate_images.py

Changed image viewer button's location to inside of accordion.
HTML file opens as window now instead of tab.

heloess committed:
Update Image Viewer.html

HTML file stops refreshing if saved.png is not found.

heloess committed:
Update sd_save_intermediate_images.py

saved.png is now saving to viewer folder

heloess committed:
Update Image Viewer.html

Now working on viewer folder

heloess committed:
viewer folder

heloess committed:
Delete scripts/Image Viewer.html

HTML file now works on viewer folder

heloess commented:
Viewer button is inside in accordion now, fixed HTML file to prevent refresh image if saved.png not found, HTML file opens as window, HTML file and saved.png is in viewer folder.

I take out my code from if statement on `def handle_image_saved` function, and now it copy final image temporarily even don't enable "Also save final image with intermediates". When click on the image viewer, it shows the final image no matter what. It works as desired.

What I'm trying to do now is to add a tick button called "Enable live preview" next to the Image viewer button, which temporarily copies the intermediate images to the viewer folder only, without relying on the "save intermediate images" button. However, since this intermediate image saving process depends on the `def callback_state(self, d):` function under the `if ssii_is_active:` statement, I could not figure out how to make it independent from the "save intermediate image" tick button.

heloess committed:
Update sd_save_intermediate_images.py

Added "Enable live preview on image viewer" button.

heloess committed:
Update sd_save_intermediate_images.py

AlUlkesh commented:
Looks good to me! Thank you. Can you add viewer\saved.png to .gitignore? Then I'll merge it.

heloess committed:
Update .gitignore

added viewer\saved.png to .gitignore
